managedplugin/download.go

```diff
@@ -52,31 +55,98 @@ func getURLLocation(ctx context.Context, org string, name string, version string
-		req, err := http.NewRequestWithContext(ctx, http.MethodHead, url, nil)
+	for _, downloadURL := range urls {
+		req, err := http.NewRequestWithContext(ctx, http.MethodHead, downloadURL, nil)
 		if err != nil {
-			return "", fmt.Errorf("failed create request %s: %w", url, err)
+			return "", fmt.Errorf("failed create request %s: %w", downloadURL, err)
 		}
 		resp, err := http.DefaultClient.Do(req)
 		if err != nil {
-			return "", fmt.Errorf("failed to get url %s: %w", url, err)
+			return "", fmt.Errorf("failed to get url %s: %w", downloadURL, err)
 		}
 		// Check server response
 		if resp.StatusCode == http.StatusNotFound {
 			resp.Body.Close()
 			continue
 		} else if resp.StatusCode != http.StatusOK {
 			resp.Body.Close()
-			fmt.Printf("Failed downloading %s with status code %d. Retrying\n", url, resp.StatusCode)
+			fmt.Printf("Failed downloading %s with status code %d. Retrying\n", downloadURL, resp.StatusCode)
 			return "", errors.New("statusCode != 200")
 		}
 		resp.Body.Close()
-		return url, nil
+		return downloadURL, nil
 	}
 
 	return "", fmt.Errorf("failed to find plugin %s/%s version %s", org, name, version)
 }
 
+func DownloadPluginFromHub(ctx context.Context, localPath string, team string, name string, version string, typ PluginType) error {
+	downloadDir := filepath.Dir(localPath)
+	if _, err := os.Stat(localPath); err == nil {
+		return nil
+	}
+
+	if err := os.MkdirAll(downloadDir, 0755); err != nil {
+		return fmt.Errorf("failed to create plugin directory %s: %w", downloadDir, err)
+	}
+
+	target := fmt.Sprintf("%s_%s", runtime.GOOS, runtime.GOARCH)
+	// We don't want to follow redirects because we want to get the download URL and show progress bar while downloading
+	client := &http.Client{
+		CheckRedirect: func(req *http.Request, via []*http.Request) error {
+			return http.ErrUseLastResponse
+		},
+	}
+	c, err := cloudquery_api.NewClient(APIBaseURL)
+	c.Client = client
+	if err != nil {
+		return fmt.Errorf("failed to create Hub API client: %w", err)
+	}
+
+	downloadURL, err := c.DownloadPluginAsset(ctx, team, cloudquery_api.PluginKind(typ.String()), name, version, target)
+	if err != nil {
+		return fmt.Errorf("failed to get plugin url: %w", err)
+	}
+	defer downloadURL.Body.Close()
+	location, ok := downloadURL.Header["Location"]
+	if !ok {
+		return fmt.Errorf("failed to get plugin url: missing location header from response")
+	}
+	if len(location) == 0 {
+		return fmt.Errorf("failed to get plugin url: empty location header from response")
+	}
+	pluginZipPath := localPath + ".zip"
+	err = downloadFile(ctx, pluginZipPath, location[0])
+	if err != nil {
+		return fmt.Errorf("failed to download plugin: %w", err)
+	}
+
+	archive, err := zip.OpenReader(pluginZipPath)
+	if err != nil {
+		return fmt.Errorf("failed to open plugin archive: %w", err)
+	}
+	defer archive.Close()
+
+	fileInArchive, err := archive.Open(fmt.Sprintf("plugin-%s-%s-%s-%s", name, version, runtime.GOOS, runtime.GOARCH))
```

Comment:
Just checking, I'm not sure: shouldn't there be some logic here for adding `.exe` for Windows?

Reply:
Good point I'll verify

Reply:
We run the tests on Windows but I'll double check

Reply:
OK so this works since we never add the `exe` on Windows when we package the zip in the SDK. We do however add it here when we unzip. So the location in the zip is the same for all OS, but we unzip to the same location.

Tests are failing since something's up with the publish and version `v3.1.10` is missing for some reason (it should be there https://github.com/cloudquery/cloudquery/actions/runs/6309661411/job/17130041346#step:8:17).

I had to use this version to get the proper binaries for each OS due to https://github.com/cloudquery/plugin-sdk/pull/1246.

That version is missing from the UI too

Reply:
Weird, I'll investigate the publishing issues. 

I wonder if we should change the publishing step to include the `.exe` inside the zip file instead 🤔 That way you can download the zip, extract and run without any renaming

Reply:
You still need to specific the filename in the zip then the output file to extract to so it doesn't really matter in that aspect (of renaming). It does matter if someone extracts the ZIP manually, and the OS tool extracts it with the name in the ZIP (without the exe)